### DISCUSSION TITLE
Adding scheduler to Nodes

### DISCUSSION THREAD

**Discussion by PerRieland** [timmbogner/Farm-Data-Relay-System#202] [Ideas]
Adding scheduler to Nodes

Timm you have everything to add scheduler to nodes. 
Here is my suggestion to add scheduling to nodes.

1: Copy “fdrs_gateway_scheduler.h” -> “fdrs_node_scheduler.h”

2: Edit “fdrs_node.h”
  Add #include "fdrs_node_scheduler.h", just below #include "fdrs_debug.h"
  Line58        #include "fdrs_debug.h"
  NEW Line   #include "fdrs_node_scheduler.h"

3: Edit “fdrs_node.h”
  Add   handle_schedule(); in void loopFDRS()
  Line 283       void loopFDRS()
                      {

  NEW Line     void loopFDRS()
                      {
                        handle_schedule();

	
After this I can use scheduleFDRS() in Nodes. 
Example in the your “Irrigation”: 
If I want the Irrigition node to report status every 30 sec. I just have to add

scheduleFDRS(checkCoils, 30000);
in setup()

You made it easy , thanks 🙂

**Comment by timmbogner** [comment]
Hi Per,
Thanks for this! The scheduler is used on the gateway to allow a user's function to be run every so often, since there is no other way to do so. On nodes, the user has control of the program, so they can set up a scheduler of their own. I may add an official scheduler for nodes, but probably only once the system needs it more.

What I'd suggest is that you create a library out of `fdrs_gateway_scheduler.h` for you to use yourself. It's a good learning experience, among other things. I actually went ahead and did the dirty work... Following is a modification of the scheduler. Copy it to a file named "cool_scheduler.h". Place the .h file in a folder entitled "cool_scheduler". Place that folder in your Arduino "libraries" folder. Now you should just need to `#include <cool_scheduler.h>` and then call `add_to_schedule()` to schedule a function and `handle_cool_schedule()` in loop.

Let me know how it works!
``` cpp


typedef struct CoolScheduleItem
{
    uint32_t interval;
    uint32_t start;
    void (*functionName)();
    bool active;
} CoolScheduleItem;

CoolScheduleItem theCoolSchedule[16];

bool add_to_schedule(void (*added_func)(), uint32_t added_interval)
{
    CoolScheduleItem arg = {.interval = added_interval, .start = millis(), .functionName = added_func, .active = true};
    for (int i = 0; i < 16; i++)
    {
        if (!theCoolSchedule[i].active)
        {
            theCoolSchedule[i] = arg;
            return true;
        }
    }
    Serial.println("Schedule is full!");
    return false;
}
void handle_cool_schedule()
{
    for (int i = 0; i < 16; i++)
    {
        if (theCoolSchedule[i].active && (millis() - theCoolSchedule[i].start > theCoolSchedule[i].interval))
        {
            theCoolSchedule[i].start = millis();
            theCoolSchedule[i].functionName();
        }
    }
}
```
